contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,397 @@
+For example, CPU time is proportioned dynamically when there is contention
+using CFS shares that ensure minimum requests are satisfied.
 
-#### QoS classes
+For incompressible resources, this prioritization scheme can inhibit the
+ability of a pod to have its requests satisfied.  For example, a Guaranteed
+pods memory request may not be satisfied if there are active BestEffort
+pods consuming all available memory.
 
-A pod can belong to one of the following 3 QoS classes: Guaranteed, Burstable, and BestEffort, in decreasing order of priority.
+The `kubelet` will support a flag `--qos-reserve-limits` that takes a
+set of percentages per compressible resource that controls how the QoS
+cgroup sandbox attempts to reserve requested resources across QoS classes
+based on the set of pods active on the node.  The flag will accept values
+in a range from 0-100%, where a value of 0 instructs the `kubelet` to
+attempt no reservation, and a value of 100 will instruct the `kubelet`
+to attempt to reserve the sum of requested resource across all pods
+on the node.  How the `kubelet` achieves this desired state is resource
+specific.  The default value per compressible resource if not specified
+is for no reservation to occur.
 
-#### Guaranteed
+The `kubelet` will allocate resources to the QoS level cgroup
+dynamically in response to the following events:
 
-`G` pods will be placed at the `$Root` cgroup by default. `$Root` is the system root i.e. "/" by default and if `--cgroup-root` flag is used then we use the specified cgroup-root as the `$Root`. To ensure Kubelet's idempotent behaviour we follow a pod cgroup naming format which is opaque and deterministic. Say we have a pod with UID: `5f9b19c9-3a30-11e6-8eea-28d2444e470d` the pod cgroup PodUID would be named: `pod-5f9b19c93a3011e6-8eea28d2444e470d`.
+1. kubelet startup/recovery
+1. pod admission
+1. pod termination
+1. at periodic intervals to reach `--qos-reserve-limits`
+heurisitcs that converge to a desired state.
 
+### QoS level CPU allocation
 
-__Note__: The cgroup-root flag would allow the user to configure the root of the QoS cgroup hierarchy. Hence cgroup-root would be redefined as the root of QoS cgroup hierarchy and not containers.
+The `BestEffort` cgroup sandbox is statically configured as follows:
 
 ```
-/PodUID/cpu.quota = cpu limit of Pod  
-/PodUID/cpu.shares = cpu request of Pod  
-/PodUID/memory.limit_in_bytes = memory limit of Pod
+ROOT/BestEffort/cpu.shares = 2
 ```
 
-Example:
+This ensures that allocation of CPU time to pods in this QoS class
+is given the lowest priority.
+
+The `Burstable` cgroup sandbox CPU share allocation is dynamic based
+on the set of pods currently scheduled to the node.  
+
+```
+ROOT/Burstable/cpu.shares = max(sum(Burstable pods cpu requests, 2)
+```
+### QoS level memory allocation
+
+By default, no memory limits are applied to the BestEffort
+and Burstable QoS level cgroups unless a `--qos-reserve-limits` value
+is specified for memory.
```

Comment:
my opinion is that if no value is provided, we should do no harm, and therefore do nothing.  it will have the same literal effect of `qos-reserve-limits=memory=0%`